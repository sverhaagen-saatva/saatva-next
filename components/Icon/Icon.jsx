import React from 'react'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import icons from './icons.json'

/**
 * Return an icon from a provided Icon List
 * @param {string} name - "name" property of icon to be found in iconList
 * @param {array} iconList - JSON Array of Objects
 * @example
 * this.getIcon('icon--star', icons.json);
 */
export function getIcon(name, iconList = icons) {
    const icon = iconList.filter(icon => icon.name === name)

    if (icon.length === 0) {
        // TODO should this throw an error in the log instead/as well?
        return false
    } else if (icon.length > 1) {
        throw new Error(`More than one icon found with the name '${name}'`)
    } else {
        return icon[0]
    }
}

/**
 * Returns Elements (shapes) in SVG
 * @param {Object} svgShapes - Object including SVG elements and nodes (paths, rects, etc.)
 * @example
 * buildSvgEls(icon.svgShapes);
 */
export function buildSvgEls(svgShapes) {
    const svgElements = Object.keys(svgShapes)
        .filter(key => svgShapes[key])
        .map(svgProp => {
            const shapes = svgShapes[svgProp]
            const Shape = `${svgProp}`

            return shapes.map((shape, index) => {
                const shapeProps = { ...shape }
                shapeProps.key = `${svgProp}${index}`

                return <Shape key={shapeProps.key} {...shapeProps} />
            })
        })

    return svgElements
}

function camelize(str) {
    return str.replace(/\W+(.)/g, function (match, chr) {
        return chr.toUpperCase();
    });
}

// stateless component!
const Icon = ({
    name,
    className,
    description,
    role,
    alt,
    style,
    width,
    height,
    fill,
    iconList,
    titleId,
    ...other
}) => {
    const icon = iconList ? getIcon(name, iconList) : getIcon(name)
    titleId = titleId ? titleId : camelize(alt)
    // setup icon className
    const buttonClasses = classNames(
        { 'icon': true },
        `icon--${name}`,
        className
    )

    const iconProps = {
        name,
        className: buttonClasses,
        role,
        alt: alt || icon.alt,
        description: description || alt,
        viewBox: icon.viewBox,
        fill,
        width: width || icon.width,
        height: height || icon.height,
        style,
        ...other
    }

    // TODO allow for passing in svgs as symbols once we have appropriate assets and webpack setup for symbols/svs
    // ex. <use xlinkHref={`#icons_${props.name}`} />
    const svgContent = icon ? buildSvgEls(icon.svgShapes) : ''

    return (
        <svg {...iconProps} aria-labelledby={titleId}>
            <title id={titleId}>{description}</title>
            {svgContent}
        </svg>
    )
}

Icon.propTypes = {
    name: PropTypes.string.isRequired,
    className: PropTypes.string,
    description: PropTypes.string,
    titleId: PropTypes.string.isRequired,
    alt: PropTypes.string.isRequired,
    role: PropTypes.string,
    viewBox: PropTypes.string,
    fill: PropTypes.string,
    width: PropTypes.string,
    height: PropTypes.string,
    style: PropTypes.object,
    iconList: PropTypes.array
}

Icon.defaultProps = {
    role: 'img',
    iconList: null
}

export default Icon
