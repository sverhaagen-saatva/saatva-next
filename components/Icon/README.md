# Icon

> **Status**: In Development

### Usage

```js
import { Icon } from 'saatva_pattern_library';
```

### Properties

| property  | propType | required | default | description
| --------  | -------- | -------- | ------- | -----------
| name | string | yes | - | 
| className | string | - | - | 
| description | string | yes | - | 
| alt | string | yes | - | 
| role | string | - | "img" | 
| viewBox | string | - | - | 
| fill | string | - | - | 
| width | string | - | - | 
| height | string | - | - | 
| style | object | -
| iconList | array | - | null

Make sure to include an 'alt' property when you add an icon to the JSON file. This will create default values for alt, ARIA-label and description.


### UX Guidance 
...


### TODOs
1. Include option to pass in symbol as the svgContent with something like:  
```jsx
<use xlinkHref={`#icons_${props.name}`} />
```
2. Add/refactor accessibility