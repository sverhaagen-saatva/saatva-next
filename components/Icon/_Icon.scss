//-------------------------------------------
// Icon Styles
// ------------------------------------------

// Icon Variables
$icon-colors-default: (
    base: $scaffold,
    article: $scaffold,
    cart: $scaffold,
    call: $scaffold,
    chat: $scaffold,
    circleCheck: $scaffold,
    close: $scaffold,
    delivery: $scaffold,
    finance: $scaffold,
    less: $scaffold,
    more: $scaffold,
    navArrow: $scaffold,
    saatvaLogo: $scaffold,
    trial: $scaffold,
    warranty: $scaffold
) !default;

$icon-colors-custom: () !default;

$icon-colors-config: map-merge($icon-colors-default, $icon-colors-custom);

.icon-- {
    &arrow-down,
    &arrow-left,
    &arrow-right,
    &bag,
    &chevron-right,
    &chevron-left,
    &call,
    &chat,
    &check,
    &close,
    &concierge,
    &cotton,
    &delivery,
    &expand,
    &fairtrade,
    &finance,
    &more,
    &less,
    &navArrow,
    &play,
    &remote,
    &removal,
    &return,
    &returns,
    &saatvaLogo,
    &staple,
    &thread,
    &trial,
    &truck,
    &warranty {
        fill: none;
        stroke: map-get($icon-colors-config, base);
        stroke-width: 1;
        stroke-linecap: round;
        stroke-linejoin: round;
    }

    @each $name, $value in $icon-colors-config {
        &#{$name} {
            stroke: $value;
        }
    }   

    &cart {
        fill: map-get($icon-colors-config, cart);
    }

    &circleCheck {
        stroke-linecap: round;
        stroke-width: 2;
        stroke: white;
        fill: none;
        background-color: map-get($icon-colors-config, circleCheck);
        border-radius: 50%;
    }

    &expand {
        transition: transform .3s ease;
    }


    &questionMark {
        circle {
            fill: map-get($icon-colors-config, base);
        }
        path {
            fill: $white;
        }
    }

    &secure {
        fill: map-get($icon-colors-config, base);
    }

    &delivery,
    &trial {
        stroke-width: 1.5;
    }
    
    &back, &close, &less, &more {
        stroke-width: 2;
    }

    &saatvaLogo {
        fill: map-get($icon-colors-config, saatvaLogo);
        stroke: none;
    }
    
    &bag, &cart {
        stroke: none;
    }

    &article {
        fill: none;
        stroke-width: 1;
        stroke-linecap: round;
        stroke-linejoin: round;

        circle {
            fill: map-get($icon-colors-config, article);
        }
        path {
            stroke: $white;
        }
    }
    &sfc {
        .st0{fill:$saatva-clay;}
        .st1{fill:none;stroke:#FFFFFF;}
        .st2{fill:$saatva-ash;}
    }
    &cottonLogo {
        .st0{fill:$saatva-clay;}
        .st1{fill:none;stroke:$saatva-clay;stroke-miterlimit:10;}
        .st2{fill:$saatva-ash;}
    }
    &certipur {
        .st0{fill:$saatva-clay;}
        .st1{fill:#FFFFFF;}
        .st2{fill:$saatva-ash;}
    }

}
