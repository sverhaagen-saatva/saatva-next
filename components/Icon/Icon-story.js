import React from 'react'
import { storiesOf } from '@storybook/react'
import { withKnobs, text, select } from '@storybook/addon-knobs/react'
// import { backgroundsSaatva } from '../../../.storybook/backgrounds'
import { withReadme } from 'storybook-readme'
import { ShowMarkup } from '../../../addons/markup'
import { Icon } from '../../components'
import iconData from '../assets/icons/icons.json'

import README from './README.md'

/* === uncomment to get correct propTypes to copy into README table === */
// import { withInfo } from '@storybook/addon-info'
// storiesOf('Icon', module).add('simple info', withInfo(`test`)(() => <Icon />))

const stories = storiesOf('Icon', module)

stories.addDecorator(withKnobs)
// stories.addDecorator(backgroundsSaatva)
stories.addDecorator(withReadme(README))

// bring in all icons from icons.json
const iconNames = iconData.map(icon => icon.name).sort()

stories
    .add('playground', () => (
        <ShowMarkup>
            <Icon
                name={select('name', iconNames, 'star')}
                className={text('className', 'new-class')}
                alt={text('alt', 'review star')}
                fill={text('fill', '#62c2b1')}
                width={text('width', '24')}
                height={text('height', '24')}
                titleId={text('titleId', '')}
            />
        </ShowMarkup>
    ))
    .add('all icons', () => {
        return (
            <div className="row">
                {
                    iconNames.map((name, i) => {
                        return (
                            <div className="col col--xs-4 col--xs-3 col--lg-2 u-marginBottom--3xl u-textCenterFlex" key={i}>
                                <Icon
                                    name={name}
                                    alt='An icon'
                                    width={text('width', '24')}
                                    height={text('height', '24')}
                                />
                                {`${name}`}
                            </div>
                        )
                    })
                }
            </div>
        )
    })
