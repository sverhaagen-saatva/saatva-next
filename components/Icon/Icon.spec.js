import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Icon from './'
import iconList from '../assets/icons/icons.json'

describe('Icon', () => {
    const props = {
        name: 'star',
        className: 'new-class',
        alt: 'review star',
        description: 'review star',
        fill: 'yellow',
        width: '24',
        height: '24',
        style: {
            opacity: '0.5'
        },
        titleId: 'review star'
    }

    const base = (
        <Icon {...props}/>
    )

    // snapshot test
    it('renders correctly', () => {
        const tree = renderer
            .create(base)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })

    const addPropsToShallowIcon = (newProps) => {
        return shallow(<Icon {...props} {...newProps} />)
    }

    // unit tests
    describe('should render correctly', () => {
        it('when a new class is added', () => {
            const wrapper = addPropsToShallowIcon({className: 'new-class'})

            expect(wrapper.hasClass('new-class')).toEqual(true)
        })
    })

    it('should accept an external icon definition list', () => {
        const wrapper = addPropsToShallowIcon({iconList: iconList})

        expect(wrapper.props().name).toEqual('star')
    })
})
