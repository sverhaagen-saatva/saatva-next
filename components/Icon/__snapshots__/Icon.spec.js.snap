// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Icon renders correctly 1`] = `
<svg
  alt="review star"
  aria-label="review star"
  className="icon icon--star new-class"
  description="review star"
  fill="yellow"
  height="24"
  name="star"
  role="img"
  style={
    Object {
      "opacity": "0.5",
    }
  }
  viewBox="0 0 17.053 16.978"
  width="24"
>
  <title>
    review star
  </title>
  <path
    d="M15.5,6.956H10.206L10.093,6.6,8.512,1.648,6.825,6.956H1.5l4.283,3.211-.118.347L3.929,15.635l4.585-3.369.3.222,4.3,3.16L11.274,9.993l.3-.216Z"
  />
</svg>
`;
